// @adiwajshing/baileys/lib/WAConnection/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WAFlag = exports.WAMetric = exports.MessageLogLevel = void 0;
var MessageLogLevel;
(function (MessageLogLevel) {
    MessageLogLevel[MessageLogLevel["none"] = 0] = "none";
    MessageLogLevel[MessageLogLevel["unhandled"] = 1] = "unhandled";
    MessageLogLevel[MessageLogLevel["all"] = 2] = "all";
})(MessageLogLevel = exports.MessageLogLevel || (exports.MessageLogLevel = {}));
var WAMetric;
(function (WAMetric) {
    WAMetric[WAMetric["liveLocation"] = 3] = "liveLocation";
    WAMetric[WAMetric["group"] = 10] = "group";
    WAMetric[WAMetric["message"] = 16] = "message";
    WAMetric[WAMetric["queryLiveLocation"] = 33] = "queryLiveLocation";
})(WAMetric = exports.WAMetric || (exports.WAMetric = {}));
var WAFlag;
(function (WAFlag) {
    WAFlag[WAFlag["ignore"] = 128] = "ignore";
    WAFlag[WAFlag["acknowledge"] = 64] = "acknowledge";
    WAFlag[WAFlag["available"] = 32] = "available";
    WAFlag[WAFlag["unavailable"] = 16] = "unavailable";
    WAFlag[WAFlag["expires"] = 8] = "expires";
    WAFlag[WAFlag["skipOffline"] = 4] = "skipOffline";
})(WAFlag = exports.WAFlag || (exports.WAFlag = {}));
