// @adiwajshing/baileys/lib/WAClient/Constants.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MimetypeMap = exports.MediaPathMap = exports.Mimetype = exports.HKDFInfoKeys = exports.WAMessageType = exports.MessageType = exports.MessageStatus = exports.Presence = void 0;
var WAMessage_1 = require("../../WAMessage/WAMessage");
/**
 * set of statuses visible to other people; see updatePresence() in WhatsAppWeb.Send
 */
var Presence;
(function (Presence) {
    Presence["available"] = "available";
    Presence["unavailable"] = "unavailable";
    Presence["composing"] = "composing";
    Presence["recording"] = "recording";
    Presence["paused"] = "paused";
})(Presence = exports.Presence || (exports.Presence = {}));
/**
 * Status of a message sent or received
 */
var MessageStatus;
(function (MessageStatus) {
    MessageStatus["sent"] = "sent";
    MessageStatus["received"] = "received";
    MessageStatus["read"] = "read";
})(MessageStatus = exports.MessageStatus || (exports.MessageStatus = {}));
/**
 * set of message types that are supported by the library
 */
var MessageType;
(function (MessageType) {
    MessageType["text"] = "conversation";
    MessageType["extendedText"] = "extendedTextMessage";
    MessageType["contact"] = "contactMessage";
    MessageType["location"] = "locationMessage";
    MessageType["liveLocation"] = "liveLocationMessage";
    MessageType["image"] = "imageMessage";
    MessageType["video"] = "videoMessage";
    MessageType["sticker"] = "stickerMessage";
    MessageType["document"] = "documentMessage";
    MessageType["audio"] = "audioMessage";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
exports.WAMessageType = function () {
    var types = WAMessage_1.proto.WebMessageInfo.WEB_MESSAGE_INFO_STUBTYPE;
    var dict = {};
    Object.keys(types).forEach(function (element) { return dict[types[element]] = element; });
    return dict;
}();
exports.HKDFInfoKeys = (_a = {},
    _a[MessageType.image] = 'WhatsApp Image Keys',
    _a[MessageType.audio] = 'WhatsApp Audio Keys',
    _a[MessageType.video] = 'WhatsApp Video Keys',
    _a[MessageType.document] = 'WhatsApp Document Keys',
    _a[MessageType.sticker] = 'WhatsApp Image Keys',
    _a);
var Mimetype;
(function (Mimetype) {
    Mimetype["jpeg"] = "image/jpeg";
    Mimetype["mp4"] = "video/mp4";
    Mimetype["gif"] = "video/gif";
    Mimetype["pdf"] = "appliction/pdf";
    Mimetype["ogg"] = "audio/ogg; codecs=opus";
    /** for stickers */
    Mimetype["webp"] = "image/webp";
})(Mimetype = exports.Mimetype || (exports.Mimetype = {}));
// path to upload the media
exports.MediaPathMap = {
    imageMessage: '/mms/image',
    videoMessage: '/mms/video',
    documentMessage: '/mms/document',
    audioMessage: '/mms/audio',
    stickerMessage: '/mms/image',
};
// gives WhatsApp info to process the media
exports.MimetypeMap = {
    imageMessage: Mimetype.jpeg,
    videoMessage: Mimetype.mp4,
    documentMessage: Mimetype.pdf,
    audioMessage: Mimetype.ogg,
    stickerMessage: Mimetype.webp,
};
